code/game/objects/items/devices/scanners.dm

```diff
@@ -7,36 +7,65 @@ GAS ANALYZER
 MASS SPECTROMETER
 REAGENT SCANNER
 */
+/obj/item/device/scanner
+	name = "scanner"
+	desc = "basic scanner unit"
+	icon_state = "multitool"
+	item_state = null
+	flags = CONDUCT
+	slot_flags = SLOT_BELT
+	w_class = ITEM_SIZE_SMALL
+	throwforce = WEAPON_FORCE_HARMLESS
+	throw_speed = 4
+	throw_range = 20
+
+	matter = list(DEFAULT_WALL_MATERIAL = 10)
+	origin_tech = null
 
+	var/obj/item/weapon/cell/cell = null
+	var/suitable_cell = /obj/item/weapon/cell/small
+
+//slime
+
+	throw_speed = 3
+	throw_range = 7
 
-/obj/item/device/healthanalyzer
+	matter = list(DEFAULT_WALL_MATERIAL = 30,"glass" = 20)
+	origin_tech = list(TECH_BIO = 1)
+
+/obj/item/device/scanner/proc/cell_check()
+	if(!cell && suitable_cell)
+		cell = new suitable_cell(src)
+
+//all user was replased on usr
+/obj/item/device/scanner/proc/cell_use_check(charge)
+	if(!cell || !cell.checked_use(charge))
+		usr << SPAN_WARNING("[src] battery is dead or missing.")
+		return
+
+/obj/item/device/scanner/New()
+	..()
+	cell_check()
+
+/obj/item/device/scanner/healthanalyzer
 	name = "health analyzer"
 	desc = "A hand-held body scanner able to distinguish vital signs of the subject."
 	icon_state = "health"
 	item_state = "analyzer"
-	flags = CONDUCT
-	slot_flags = SLOT_BELT
-	throwforce = 3
-	w_class = ITEM_SIZE_SMALL
 	throw_speed = 5
 	throw_range = 10
+
 	matter = list(DEFAULT_WALL_MATERIAL = 200)
 	origin_tech = list(TECH_MAGNET = 1, TECH_BIO = 1)
-	var/mode = 1
-	var/obj/item/weapon/cell/cell = null
-	var/suitable_cell = /obj/item/weapon/cell/small
 
-/obj/item/device/healthanalyzer/New()
-	..()
-	if(!cell && suitable_cell)
-		cell = new suitable_cell(src)
+	var/mode = 1
 
-/obj/item/device/healthanalyzer/attack(mob/living/M, mob/living/user)
-	if(!cell || !cell.checked_use(3))
-		user << SPAN_WARNING("[src] battery is dead or missing.")
+/obj/item/device/scanner/healthanalyzer/attack(mob/living/M, mob/living/user)
+	cell_use_check(3)
```

Comment:
This won't cancel out a scan if there's a dead or missing battery, it'll just display the "It's out of battery" warning and let the scan go through anyhow.

Replace it with if(cell_use_check(3)), and make cell_use_check() return either TRUE or FALSE depending on if the check failed or not.

You could also move this check, along with the user.incapacitated() check to the /obj/item/device/scanner type, and call it in the subtype's attack procs with
`. = ..()
if(!.)
    return`